fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -1270,18 +1353,129 @@ ACTOR Future<Void> servicePeekRequest(
+		storageTeamData->poppedRecently = true;
+
+		if (storageTeamData->unpoppedRecovered && upTo > logData->recoveredAt) {
+			storageTeamData->unpoppedRecovered = false;
+			logData->unpoppedRecoveredStorageTeams--;
+			TraceEvent("TLogPoppedStorageTeam", logData->logId)
+			    .detail("Teamss", logData->unpoppedRecoveredStorageTeams)
+			    .detail("DurableKCVer", logData->durableKnownCommittedVersion)
+			    .detail("RecoveredAt", logData->recoveredAt);
+			if (logData->unpoppedRecoveredStorageTeams == 0 &&
+			    logData->durableKnownCommittedVersion >= logData->recoveredAt && logData->recoveryComplete.canBeSet()) {
+				logData->recoveryComplete.send(Void());
+			}
+		}
+
+		uint64_t PoppedVersionLag = logData->persistentDataDurableVersion - logData->queuePoppedVersion;
+		if (SERVER_KNOBS->ENABLE_DETAILED_TLOG_POP_TRACE &&
+		    (logData->queuePoppedVersion > 0) && // avoid generating massive events at beginning
+		    (storageTeamData->unpoppedRecovered ||
+		     PoppedVersionLag >=
+		         SERVER_KNOBS->TLOG_POPPED_VER_LAG_THRESHOLD_FOR_TLOGPOP_TRACE)) { // when recovery or long lag
+			TraceEvent("TLogPopDetails", logData->logId)
+			    .detail("StorageTeam", storageTeamData->storageTeamId)
+			    .detail("UpTo", upTo)
+			    .detail("PoppedVersionLag", PoppedVersionLag)
+			    // TODO: add minPopStorageTeam after having popDiskQueue()
+			    .detail("QueuePoppedVersion", logData->queuePoppedVersion)
+			    .detail("UnpoppedRecovered", storageTeamData->unpoppedRecovered ? "True" : "False")
+			    .detail("NothingPersistent", storageTeamData->nothingPersistent ? "True" : "False");
+		}
+
+		Version minVersionInTeam = upTo; // storageTeams
+		std::vector<Tag> allTags = logData->storageTeams[storageTeamID];
+		for (const Tag& tag : allTags) {
+			// find the lowest version of all tags in this storage team
+			if (storageTeamData->poppedTagVersions.find(tag) == storageTeamData->poppedTagVersions.end()) {
+				// question : what if a tag has never been popped, which version can we use to pop?
+				minVersionInTeam = std::numeric_limits<int>::max();
+				break;
+			}
+			minVersionInTeam = std::min(minVersionInTeam, storageTeamData->poppedTagVersions[tag]);
+		}
+
+		storageTeamData->popped = minVersionInTeam;
+		// pop from in-memory object: TagData::versionMessages
+		// only when minVersionInTeam > logData->persistentDataDurableVersion, there are data
+		// need to be popped from in memory data structure(i.e. versionMessages) that has not been persisted.
+		// if minVersionInTeam < logData->persistentDataDurableVersion, it means all the data needs to be popped has
+		// been persisted into disk, so they are already erased from in memory data structure in updatePersistentData(),
+		// thus we only need to erase them from disk.
+		if (minVersionInTeam > logData->persistentDataDurableVersion)
+			wait(storageTeamData->eraseMessagesBefore(minVersionInTeam, self, logData, TaskPriority::TLogPop));
```

Comment:
Is popping spilled data implemented? Looks like this PR also doesn't address this.

Reply:
that is correct, anything on the disk is not being erased now, as I said in the description, `This commit only changes the memory part, disk part will be a separate PR`